components/tidb_query_common/src/storage/scanner.rs

```diff
@@ -23,6 +33,34 @@ pub struct RangesScanner<T> {
     current_range: IntervalRange,
     working_range_begin_key: Vec<u8>,
     working_range_end_key: Vec<u8>,
+    rescheduler: RescheduleChecker,
+}
+
+// TODO: maybe it's better to make it generic to avoid directly depending
+// on yatp's rescheduler.
+struct RescheduleChecker {
+    prev_start: Instant,
+    prev_key_count: usize,
+}
+
+impl RescheduleChecker {
+    fn new() -> Self {
+        Self {
+            prev_start: Instant::now(),
+            prev_key_count: 0,
+        }
+    }
+
+    async fn check_reschedule(&mut self, force_check: bool) {
+        self.prev_key_count += 1;
+        if (force_check || self.prev_key_count % CHECK_KEYS == 0)
+            && self.prev_start.saturating_elapsed() > MAX_TIME_SLICE
+        {
+            reschedule().await;
+            self.prev_start = Instant::now();
```

Comment:
Is this instant using Coarse timer? We may need to make sure of it, as now it is called pretty frequently, once every 32 rows.

Reply:
In my benchmark (in the internal idc linux environment),  one call of `now` cost 28ns while `now_coarse` is 7ns. Not sure there are big difference in other env(like arm linux?). As the precision of `_coarse` is about 1ms, it a bit too large because the time slice is also 1ms. But in my test, there is no observable performance different between this two method.